rand_core/src/lib.rs

```diff
@@ -247,55 +247,136 @@ pub enum ErrorKind {
+enum ErrorDetails {
+    None,
+    Str(&'static str),
+    #[cfg(feature="std")]
+    Error(Box<std::error::Error + Send + Sync>),
 }
 
-#[cfg(not(feature="std"))]
-#[derive(Debug, Clone, PartialEq, Eq)]
+impl fmt::Display for ErrorDetails {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use ::std::error::Error;
+        write!(f, "{}", self.description())
+    }
+}
+
+#[cfg(feature="std")]
+impl ::std::error::Error for ErrorDetails {
+    fn description(&self) -> &str {
+        match *self {
+            ErrorDetails::Str(ref s) => s,
+            _ => unreachable!(),
+        }
+    }
+    
+    // no cause: Error case gets unwrapped by impl for Error
+}
+
+/// Error type of random number generators
+/// 
+/// This embeds two things, an `ErrorKind`, which can be matched over, and
+/// optional details provided to `new_str` or `new_err`.
+#[derive(Debug)]
 pub struct Error {
+    /// Error kind
     pub kind: ErrorKind,
-    pub cause: Option<&'static str>,
+    details: ErrorDetails,
 }
 
 impl Error {
+    /// Create a new instance, with no details.
+    pub fn new(kind: ErrorKind) -> Self {
+        Self { kind, details: ErrorDetails::None }
+    }
+    
+    /// Create a new instance, with a static string describing the details.
+    /// 
+    /// This may be used in both `std` and `no_std` enviroments.
+    pub fn new_str(kind: ErrorKind, details: &'static str) -> Self {
+        Self { kind, details: ErrorDetails::Str(details) }
+    }
+    
+    /// Create a new instance, with a chained `Error` cause.
     #[cfg(feature="std")]
-    pub fn new(kind: ErrorKind, cause: Option<Box<std::error::Error>>) -> Error {
-        Error {
-            kind: kind,
-            cause: cause,
+    pub fn new_err<E>(kind: ErrorKind, cause: E) -> Self
+        where E: Into<Box<std::error::Error + Send + Sync>>
+    {
+        Self { kind, details: ErrorDetails::Error(cause.into()) }
+    }
+    
+    /// Get details on the error, if available (string message or chained
+    /// "cause").
+    /// 
+    /// In the case of a chained error, the actual error can be obtained with
+    /// `std::error::Error::cause`.
+    pub fn details(&self) -> Option<&str> {
+        match self.details {
+            ErrorDetails::None => None,
+            ErrorDetails::Str(ref s) => Some(s),
+            ErrorDetails::Error(ref e) => Some(e.description()),
         }
     }
 }
 
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self.kind {
-            ErrorKind::Unavailable => write!(f, "RNG not available."),
-            ErrorKind::Transient => write!(f, "RNG has failed, probably temporary."),
-            ErrorKind::NotReady => write!(f, "RNG not ready yet."),
-            ErrorKind::Other => write!(f, "An unspecified RNG error occurred."),
-            ErrorKind::__Nonexhaustive => unreachable!(),
-        }
+        write!(f, "RNG error: {}", self.kind.description())
```

Comment:
According to the [book](https://doc.rust-lang.org/book/first-edition/error-handling.html#the-error-trait) `description` is supposed to be a short description, and `Display` a user-facing representation of the error (a proper sentence with capital letter, dot at the end, and a bit less condensed text).

I think this is okay, but like the previous version more.